Assign course/script to newly created section

Bjvanminnen opened the pull request:
This PR makes it so that when we click on New Section in the assign course dropdown, we load the page with the section edit experience started. This should be true whether our experiment is on or off.
![image](https://user-images.githubusercontent.com/1767466/29291033-181d85e8-80f7-11e7-87d0-889bb8987795.png)

It does so by going to the page with courseId/scriptId query params. We have a redux action that will parse these query params, and put them in the store. Not sure if it would be better to keep the query param parse logic outside of redux?

Bjvanminnen commented on apps/src/templates/teacherDashboard/teacherSectionsRedux.js:
```diff
@@ -93,6 +95,34 @@ export const setValidAssignments = (validCourses, validScripts) => ({
+ */
+export function getQueryString_() {
+  return window.location.search;
+}
```
Don't know if there's a better way to be able to stub this, but this seems to work (semi-copied from what we do in experiments.js)

islemaster replied:
Seems fine to me.

islemaster commented on apps/src/sites/code.org/pages/public/teacher-dashboard/sections.js:
```diff
@@ -32,25 +32,14 @@ export function renderSectionsPage(data) {
-    defaultScriptId = parseInt(query.scriptId, 10);
-  }
+  store.dispatch(setDefaultAssignment());
+  console.log('dispatched');
```
Leftover `console.log`

islemaster commented on apps/src/templates/courseOverview/AssignToSection.js:
```diff
@@ -78,6 +79,14 @@ class AssignToSection extends Component {
+    if (experiments.isEnabled(SECTION_FLOW_2017)) {
+      hrefNewSection = '/home?' + queryString.stringify({courseId, scriptId});
+    } else {
+      hrefNewSection = `${window.dashboard.CODE_ORG_URL}/teacher-dashboard?` +
```
Could use the `pegasus()` helper here.

islemaster commented on apps/src/templates/teacherDashboard/teacherSectionsRedux.js:
```diff
@@ -382,6 +414,15 @@ export default function teacherSections(state=initialState, action) {
+      ...state,
+      defaultCourseId: courseId,
+      defaultScriptId: scriptId
+    };
```
I'm confused - don't we want to launch straight into the new section flow when we pass these queryparams?  Why wouldn't we dispatch _that_ and set the course and script ids on the `sectionBeingEdited` instead of introducing some new and independent state?

Bjvanminnen replied:
Hmm, that might make more sense. Let me give that a try.

Bjvanminnen commented:
Your approach makes a lot more sense. I've updated to do that. As part of this, I moved all of the query param stuff out of redux (where it didn't really feel right). Though it's semi-duplicated between the two entry points, one of these will go away once we don't have the experiment.

Bjvanminnen commented on apps/src/templates/teacherDashboard/OwnedSections.jsx:
```diff
@@ -40,6 +38,8 @@ class OwnedSections extends React.Component {
     beginEditingSection: PropTypes.func.isRequired,
     beginImportRosterFlow: PropTypes.func.isRequired,
+    defaultCourseId: PropTypes.number,
+    defaultScriptId: PropTypes.number,
```
TODO: remove these props

Bjvanminnen commented on apps/test/unit/templates/teacherDashboard/teacherSectionsReduxTest.js:
```diff
@@ -1,6 +1,7 @@
 import sinon from 'sinon';
 import { assert, expect } from '../../../util/configuredChai';
 import {stubRedux, restoreRedux, registerReducers, getStore} from '@cdo/apps/redux';
+// Use wildcard import so that we can stub
```
remove comment

Bjvanminnen commented:
Yah UI tests for catching that

Bjvanminnen commented:
@islemaster If you could take another look at this PR and sign off if things look good, that would be great :)

islemaster commented on apps/src/templates/teacherDashboard/OwnedSections.jsx:
```diff
@@ -70,13 +60,12 @@ class OwnedSections extends React.Component {
+      this.props.beginEditingNewSection();
     } else {
       // This is the only usage of the newSection action, and can be removed once
       // SECTION_FLOW_2017 is finished
```
Note: This comment is not longer accurate :)

islemaster commented on apps/src/templates/teacherDashboard/OwnedSections.jsx:
```diff
@@ -70,13 +60,12 @@ class OwnedSections extends React.Component {
       // This is the only usage of the newSection action, and can be removed once
       // SECTION_FLOW_2017 is finished
-      return this.props.newSection(defaultCourseId);
+      return this.props.newSection();
```
Not blocking this PR (this is really my bad) but note to self: We should have a unit test for OwnedSections.jsx that clicks this button both with the experiment on and off.
